### PR TITLE
[WIP] Add RESTful API for mitmweb option

### PR DESCRIPTION
**Pull request by MatthewShao**
- [x] Implement GET /options
- [x] Implement POST /options
The return json looks like this:
```python
[
    {
        "default": false, 
        "help": "Add all certificates of the upstream server to the certificate chain that will be served to the proxy client, as extras.", 
        "name": "add_upstream_certs_to_client_chain", 
        "type": "bool", 
        "value": false
    }, 
    {
        "default": false, 
        "help": "Strip out request headers that might cause the server to return 304-not-modified.", 
        "name": "anticache", 
        "type": "bool", 
        "value": false
    }, 
...]
```
The tests passed on my local repo, but kept falling on Travis... Is that because of some compatibility problem?

### CODE REVIEW

File: mitmproxy/optmanager.py [hunk 1]
```diff
@@ -416,6 +416,20 @@ def dump_defaults(opts):
     return ruamel.yaml.round_trip_dump(s)
 
 
+def dump_dicts(opts):
+    """
+        Dumps the options into a list of dict object.
+
+        Return: A list like: [ { name: "anticahce", type: "bool", default: false, value: true, help: "help text"}]
```
Comment (mhils) [inline mitmproxy/optmanager.py]: typo: anticache

File: mitmproxy/optmanager.py [hunk 2]
```diff
@@ -416,6 +416,20 @@ def dump_defaults(opts):
...
+    """
+    options_list = []
+    for k in sorted(opts.keys()):
+        o = opts._options[k]
```
Comment (mhils) [inline mitmproxy/optmanager.py]: You can simplify this a tiny bit:
```python
for k, o in sorted(opts.items()):
```

File: mitmproxy/optmanager.py [hunk 3]
```diff
@@ -416,6 +416,20 @@ def dump_defaults(opts):
...
+        option = {'name': k, 'type': o.typespec.__name__, 'default': o.default, 'value': o.current(), 'help': o.help.strip()}
```
Comment (mhils) [inline mitmproxy/optmanager.py]: Maybe add some line breaks here to make this easier to read?

Comment (cortesi) [inline mitmproxy/optmanager.py]: Also, the help.strip() here should be un-necessary - we dedent and format help text on initialisation. Let's keep the help formatting in one place.

Comment (mhils) [inline mitmproxy/optmanager.py]: Good catch. This also needs to support `.choices`.

File: mitmproxy/optmanager.py [hunk 4]
```diff
@@ -416,6 +416,35 @@ def dump_defaults(opts):
...
+def dump_dicts(opts):
+    """
+        Dumps the options into a list of dict object.
+
+        Return: A list like: [ { name: "anticache", type: "bool", default: false, value: true, help: "help text"} ]
+    """
+    options_list = []
+    for k in sorted(opts.keys()):
+        o = opts._options[k]
+        if o.typespec in (str, int, bool):
```
Comment (mhils) [inline mitmproxy/optmanager.py]: Move this typespec to string conversion into its own function in mitmproxy/utils/typecheck? :)

File: mitmproxy/optmanager.py (hunk 2 above)
Comment (MatthewShao) [inline mitmproxy/optmanager.py]: I'm afraid we could not do this, `opts` here is not a `dict`, but a `OptManager`, or do you think we should add a `.items` method for it?

Comment (mhils) [inline mitmproxy/optmanager.py]: Ah, I see. Let's keep it as is then. :)

File: mitmproxy/optmanager.py (hunk 4 above)
Comment (MatthewShao) [inline mitmproxy/optmanager.py]: Good advice!

File: test/mitmproxy/tools/web/test_app.py [hunk 5]
```diff
@@ -253,6 +253,15 @@ def test_settings_update(self):
...
         assert self.put_json("/settings", {"wtf": True}).code == 400
 
+    def test_options(self):
+        j = json(self.fetch("/options"))
+        assert type(j) == list
+        assert type(j[0]) == dict
+
+    def test_option_update(self):
+        assert self.put_json("/options", {"anticache": True}).code == 200
+        assert self.put_json("/options", {"wtf": True}).code == 400
```
Comment (mhils) [inline test/mitmproxy/tools/web/test_app.py]: We should also test type confusion here (e.g. `{"anticache": "foo"}`)

File: mitmproxy/utils/typecheck.py [hunk 6]
```diff
@@ -98,3 +98,15 @@ def check_option_type(name: str, value: typing.Any, typeinfo: typing.Any) -> Non
         return
     elif not isinstance(value, typeinfo):
         raise e
+
+
+def typespec_to_str(typespec: typing.Any) -> str:
```
Comment (mhils) [inline mitmproxy/utils/typecheck.py]: We should also use this in `dump_defaults`

File: mitmproxy/utils/typecheck.py [hunk 7]
```diff
@@ -98,3 +98,15 @@ def check_option_type(name: str, value: typing.Any, typeinfo: typing.Any) -> Non
...
+    if typespec in (str, int, bool):
+        t = typespec.__name__
+    elif typespec == typing.Optional[str]:
+        t = 'Union'
```
Comment (mhils) [inline mitmproxy/utils/typecheck.py]: Make this `"optional str"` (and `"sequence of str"` below this) to match what is currently being done in `dump_defaults`

**Pull request merged**